File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    static int lessThanEquals(int[] a, int k) {
        if(k<0) return 0;
        int sum = 0;
        int count = 0;
        int left = 0;
        int right = 0;
        while(right<a.length){
            sum = sum + a[right];
            while(sum > k){
                sum = sum - a[left];
                left++;
            }
            count = count + right - left + 1;
            right++;
        }
        return count;
    }
    public int numSubarraysWithSum(int[] a, int k) {
        return lessThanEquals(a,k) - lessThanEquals(a,k-1);
    }
}

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.go
func lessThanEquals(a []int, k int) int {
    if (k < 0) {
        return 0
    }
    sum := 0
    count := 0
    left := 0
    right := 0
    for (right < len(a)) {
        sum += a[right]
        for (sum > k) {
            sum -= a[left]
            left++
        }
        count += right - left + 1
        right++
    }
    return count
}

func numSubarraysWithSum(nums []int, goal int) int {
    return lessThanEquals(nums, goal) - lessThanEquals(nums, goal-1)
}